python_modules/dagster/dagster/config/iterate_types.py

```diff
@@ -1,27 +1,32 @@
+from typing import Dict, Literal, Union, cast
 from dagster import check
+from dagster.config.field import Field
 
-from .config_type import ConfigType, ConfigTypeKind
+from .config_type import Array, Noneable, ConfigType, ConfigTypeKind, ScalarUnion
 from .snap import ConfigSchemaSnapshot, snap_from_config_type
 
 
-def iterate_config_types(config_type):
+def iterate_config_types(config_type: ConfigType):
     check.inst_param(config_type, "config_type", ConfigType)
-    if config_type.kind == ConfigTypeKind.ARRAY or config_type.kind == ConfigTypeKind.NONEABLE:
+    if isinstance(config_type, (Array, Noneable)):
         yield from iterate_config_types(config_type.inner_type)
 
     if ConfigTypeKind.has_fields(config_type.kind):
-        for field_type in config_type.fields.values():
-            yield from iterate_config_types(field_type.config_type)
+        fields = cast(Dict[str, Field], config_type.fields)  # type: ignore
+        for field in fields.values():
+            yield from iterate_config_types(field.config_type)
 
-    if config_type.kind == ConfigTypeKind.SCALAR_UNION:
+    if isinstance(config_type, ScalarUnion):
```

Comment:
I feel a little weary about this change as mentioned https://app.graphite.dev/github/pr/dagster-io/dagster/5939. Think I would prefer dropping the type annotation here for now and leaving a comment about how the `kind` setup conflicts with `mypy`